astropy/coordinates/builtin_frames/icrs_cirs_transforms.py

```diff
@@ -113,27 +115,36 @@ def cirs_to_icrs(cirs_coo, icrs_frame):
 
 @frame_transform_graph.transform(FunctionTransformWithFiniteDifference, ICRS, GCRS)
 def icrs_to_gcrs(icrs_coo, gcrs_frame):
-    # first set up the astrometry context for ICRS<->GCRS.
+    # Set up the astrometry context for ICRS <-> GCRS.
     astrom = erfa_astrom.get().apcs(gcrs_frame)
 
-    if (
-        icrs_coo.data.name == "unitspherical"
-        or icrs_coo.data.to_cartesian().x.unit == u.one
-    ):
-        # if no distance, just do the infinite-distance/no parallax calculation
+    # Handle infinite distances properly
+    if np.isinf(icrs_coo.distance):
+        icrs_coo = icrs_coo.realize_frame(
+            SphericalRepresentation(icrs_coo.ra, icrs_coo.dec, 1e12 * u.kpc)  # Large number
+        )
+
+    if icrs_coo.data.name == "unitspherical" or icrs_coo.data.to_cartesian().x.unit == u.one:
+        if not get_inf_dist():
+            raise ValueError(
+                "Transformation requires a distance. "
+                "Use `set_inf_dist()` context manager to override."
+            )
+        
+        # No parallax calculation for infinite distance
         srepr = icrs_coo.represent_as(SphericalRepresentation)
         gcrs_ra, gcrs_dec = atciqz(srepr.without_differentials(), astrom)
 
+        if np.isnan(gcrs_ra) or np.isnan(gcrs_dec):
+            print("Warning: NaN detected in icrs_to_gcrs transformation!")
+
```

Comment:
Why did you add these NaN checks as part of this PR?  I'd remove them if it's not related.  Regardless, the message would need to be improved, and it should not be emitted using `print()`.